examples/audio/lhotse_example.py

```diff
@@ -0,0 +1,235 @@
+
+    def _collect_batch(self):
+        self.time_constraint.reset()
+        cuts = []
+        while True:
+            # Check that we have not reached the end of the dataset.
+            try:
+                if self.reuse_cuts_buffer:
+                    next_cut = self.reuse_cuts_buffer.popleft()
+                else:
+                    # If this doesn't raise (typical case), it's not the end: keep processing.
+                    next_cut = next(self.datapipe)
+            except StopIteration:
+                # No more cuts to sample from: if we have a partial batch,
+                # we may output it, unless the user requested to drop it.
+                # We also check if the batch is "almost there" to override drop_last.
+                if cuts and (
+                    not self.drop_last or self.time_constraint.close_to_exceeding()
+                ):
+                    # We have a partial batch and we can return it.
+                    self.diagnostics.keep(cuts)
+                    return CutSet.from_cuts(cuts)
+                else:
+                    # There is nothing more to return or it's discarded:
+                    # signal the iteration code to stop.
+                    self.diagnostics.discard(cuts)
+                    raise StopIteration()
+
+            # Track the duration/frames/etc. constraints.
+            self.time_constraint.add(next_cut)
+            next_num_cuts = len(cuts) + 1
+
+            # Did we exceed the max_frames and max_cuts constraints?
+            if not self.time_constraint.exceeded() and (
+                self.max_cuts is None or next_num_cuts <= self.max_cuts
+            ):
+                # No - add the next cut to the batch, and keep trying.
+                cuts.append(next_cut)
+            else:
+                # Yes. Do we have at least one cut in the batch?
+                if cuts:
+                    # Yes. Return the batch, but keep the currently drawn cut for later.
+                    self.reuse_cuts_buffer.append(next_cut)
+                    break
+                else:
+                    # No. We'll warn the user that the constrains might be too tight,
+                    # and return the cut anyway.
+                    warnings.warn(
+                        "The first cut drawn in batch collection violates "
+                        "the max_frames, max_cuts, or max_duration constraints - "
+                        "we'll return it anyway. "
+                        "Consider increasing max_frames/max_cuts/max_duration."
+                    )
+                    cuts.append(next_cut)
+
+        self.diagnostics.keep(cuts)
+        return CutSet.from_cuts(cuts)
+
+
+class IODataPipe(dp.IterDataPipe):
+    def __init__(self, datapipe):
+        self.datapipe = datapipe
+
+    def __iter__(self):
+        for cut_idx, batch_idx, batch_size, cut in self.datapipe:
+            yield cut_idx, batch_idx, batch_size, cut.load_audio(), cut
```

Comment:
Could you use `Mapper` to replace this DataPipe?
You should be able to use functional API as:
```py
source_datapipe # Referring to self.datapipe
source_datapipe.map(fn=lambda x: x.load_audio(), input_col=3, output_col=-1)
```
The result is going to be `cut_idx, batch_idx, batch_size, cut, cut.load_audio()`

Reply:
This looks interesting. Will the lambda work well with pickling / multiprocessing though?

Reply:
Also: is it possible to work with dicts rather than tuples when using the functional API? E.g.:

```python
yield {
  'cut_idx': cut_idx,
  'batch_idx': batch_idx,
  ...
}
```

Reply:
> Will the lambda work well with pickling / multiprocessing though?

We want to support `lambda` as it's a common feature for functional programming. So, as long as `dill` is installed, we would let multiprocessing with lambda works.

> Also: is it possible to work with dicts rather than tuples when using the functional API?

Yes. It's supposed to work if you give `input_col='cut_idx'`.